File: Core/services/document_service.py
"""
Document Service — Document Management & Processing.

Responsibility: Index, retrieve, and manage documents.
  - Document indexing (text and files)
  - Text extraction
  - Metadata management

Pure Python, type hinted, fully testable.
Adapted from RAG_RAT/Core/services/document_service.py.
"""

from __future__ import annotations

import json
import logging
import uuid
from datetime import datetime
from pathlib import Path
from typing import Any, Dict, List, Optional

from Core.exceptions import DocumentError, ValidationError

logger = logging.getLogger(__name__)


class DocumentService:
    """
    Service for document management.

    Pure business logic — no UI dependencies.
    """

    SUPPORTED_FORMATS = {
        ".txt": "text/plain",
        ".pdf": "application/pdf",
        ".md": "text/markdown",
        ".json": "application/json",
        ".csv": "text/csv",
        ".py": "text/x-python",
        ".html": "text/html",
        ".xml": "text/xml",
        ".rst": "text/x-rst",
        ".rtf": "application/rtf",
    }

    MAX_FILE_SIZE = 100 * 1024 * 1024  # 100 MB

    def __init__(self, storage_dir: Optional[Path] = None):
        if storage_dir is None:
            try:
                from config_system import config

                storage_dir = config.BASE_DIR / "rag_storage"
            except ImportError:
                storage_dir = Path.cwd() / "rag_storage"
        self.storage_dir = Path(storage_dir)
        self.storage_dir.mkdir(parents=True, exist_ok=True)
        self.metadata_file = self.storage_dir / ".metadata.json"
        logger.info(f"✓ DocumentService initialized: {self.storage_dir}")

    # ─── Index text ──────────────────────────────────────

    def index_text(
        self,
        text: str,
        source: str,
        metadata: Optional[Dict[str, Any]] = None,
    ) -> str:
        """
        Index raw text content.

        Returns:
            Document ID.
        """
        if not text or not text.strip():
            raise ValidationError("Text cannot be empty", field="text")
        if not source or not source.strip():
            raise ValidationError("Source cannot be empty", field="source")
        if len(text) > 10 * 1024 * 1024:
            raise ValidationError("Text too large (max 10 MB)", field="text")

        try:
            doc_id = self._generate_id()
            doc_meta = {
                "id": doc_id,
                "name": f"{source}_{doc_id}",
                "source": source,
                "format": "text",
                "size": len(text),
                "created_at": datetime.now().isoformat(),
                "indexed": True,
            }
            if metadata:
                doc_meta.update(metadata)

            content_path = self.storage_dir / f"{doc_id}.txt"
            content_path.write_text(text, encoding="utf-8")
            self._save_metadata(doc_id, doc_meta)

            logger.info(f"✓ Indexed text: {doc_id} ({len(text)} chars)")
            return doc_id
        except (ValidationError, DocumentError):
            raise
        except Exception as exc:
            raise DocumentError(f"Failed to index text: {exc}", file_path=source)

    # ─── Index file ──────────────────────────────────────

    def index_file(
        self,
        file_path: str,
        metadata: Optional[Dict[str, Any]] = None,
    ) -> str:
        """
        Index a file.

        Returns:
            Document ID.
        """
        fp = Path(file_path)
        if not fp.exists():
            raise DocumentError(f"File not found: {fp}", file_path=str(fp))
        if fp.suffix.lower() not in self.SUPPORTED_FORMATS:
            raise DocumentError(
                f"Unsupported format: {fp.suffix}",
                file_path=str(fp),
                format=fp.suffix,
            )
        if fp.stat().st_size > self.MAX_FILE_SIZE:
            raise ValidationError(
                f"File too large ({fp.stat().st_size} bytes)",
                field="file_size",
            )

        try:
            text = self.extract_text(str(fp))
            return self.index_text(
                text,
                source=fp.name,
                metadata={
                    "original_path": str(fp),
                    "file_format": fp.suffix,
                    **(metadata or {}),
                },
            )
        except (ValidationError, DocumentError):
            raise
        except Exception as exc:
            raise DocumentError(f"Failed to index file: {exc}", file_path=str(fp))

    # ─── Text extraction ────────────────────────────────

    def extract_text(self, file_path: str) -> str:
        """Extract text from a file."""
        fp = Path(file_path)
        suffix = fp.suffix.lower()

        try:
            if suffix == ".pdf":
                return self._extract_pdf(fp)
            return fp.read_text(encoding="utf-8", errors="ignore")
        except Exception as exc:
            raise DocumentError(f"Extraction failed: {exc}", file_path=str(fp))

    # ─── Listing ─────────────────────────────────────────

    def list_indexed_documents(self) -> List[Dict[str, Any]]:
        """List all indexed documents."""
        all_meta = self._load_all_metadata()
        return list(all_meta.values())

    # ─── Private helpers ─────────────────────────────────

    @staticmethod
    def _generate_id() -> str:
        return str(uuid.uuid4())[:8]

    @staticmethod
    def _extract_pdf(path: Path) -> str:
        """Extract text from PDF using PyMuPDF (fitz) if available."""
        try:
            import fitz  # PyMuPDF

            doc = fitz.open(str(path))
            pages = [page.get_text() for page in doc]
            doc.close()
            return "\n\n".join(pages)
        except ImportError:
            try:
                from pypdf import PdfReader

                reader = PdfReader(str(path))
                return "\n\n".join(p.extract_text() or "" for p in reader.pages)
            except ImportError:
                raise DocumentError(
                    "No PDF library available (install PyMuPDF or pypdf)",
                    file_path=str(path),
                    format=".pdf",
                )

    def _save_metadata(self, doc_id: str, meta: Dict[str, Any]) -> None:
        all_meta = self._load_all_metadata()
        all_meta[doc_id] = meta
        self.metadata_file.write_text(json.dumps(all_meta, indent=2), encoding="utf-8")

    def _load_all_metadata(self) -> Dict[str, Any]:
        if self.metadata_file.exists():
            try:
                return json.loads(self.metadata_file.read_text(encoding="utf-8"))
            except Exception:
                return {}
        return {}

File: rust_output/src/document_service.rs
/// Document Service — Document Management & Processing.
/// 
/// Responsibility: Index, retrieve, and manage documents.
/// - Document indexing (text and files)
/// - Text extraction
/// - Metadata management
/// 
/// Pure Python, type hinted, fully testable.
/// Adapted from RAG_RAT/Core/services/document_service::py.

use anyhow::{Result, Context};
use crate::exceptions::{DocumentError, ValidationError};
use std::collections::HashMap;
use std::fs::File;
use std::io::{self, Read, Write};
use std::path::PathBuf;

pub static LOGGER: std::sync::LazyLock<String /* logging::getLogger */> = std::sync::LazyLock::new(|| Default::default());

/// Service for document management.
/// 
/// Pure business logic — no UI dependencies.
#[derive(Debug, Clone)]
pub struct DocumentService {
    pub storage_dir: PathBuf,
    pub metadata_file: String,
}

impl DocumentService {
    pub fn new(storage_dir: Option<PathBuf>) -> Self {
        Self {
            storage_dir: PathBuf::from(storage_dir),
            metadata_file: (self.storage_dir / ".metadata.json".to_string()),
        }
    }
    /// Index raw text content.
    /// 
    /// Returns:
    /// Document ID.
    pub fn index_text(&mut self, text: String, source: String, metadata: Option<HashMap<String, Box<dyn std::any::Any>>>) -> Result<String> {
        // Index raw text content.
        // 
        // Returns:
        // Document ID.
        if (!text || !text.trim().to_string()) {
            return Err(anyhow::anyhow!("ValidationError('Text cannot be empty', field='text')"));
        }
        if (!source || !source.trim().to_string()) {
            return Err(anyhow::anyhow!("ValidationError('Source cannot be empty', field='source')"));
        }
        if text.len() > ((10 * 1024) * 1024) {
            return Err(anyhow::anyhow!("ValidationError('Text too large (max 10 MB)', field='text')"));
        }
        // try:
        {
            let mut doc_id = self._generate_id();
            let mut doc_meta = HashMap::from([("id".to_string(), doc_id), ("name".to_string(), format!("{}_{}", source, doc_id)), ("source".to_string(), source), ("format".to_string(), "text".to_string()), ("size".to_string(), text.len()), ("created_at".to_string(), datetime::now().isoformat()), ("indexed".to_string(), true)]);
            if metadata {
                doc_meta.extend(metadata);
            }
            let mut content_path = (self.storage_dir / format!("{}.txt", doc_id));
            content_pathstd::fs::write(&text));
            self._save_metadata(doc_id, doc_meta);
            logger.info(format!("✓ Indexed text: {} ({} chars)", doc_id, text.len()));
            doc_id
        }
        // except (ValidationError, DocumentError) as _e:
        // except Exception as exc:
    }
    /// Index a file.
    /// 
    /// Returns:
    /// Document ID.
    pub fn index_file(&mut self, file_path: String, metadata: Option<HashMap<String, Box<dyn std::any::Any>>>) -> Result<String> {
        // Index a file.
        // 
        // Returns:
        // Document ID.
        let mut fp = PathBuf::from(file_path);
        if !fp.exists() {
            return Err(anyhow::anyhow!("DocumentError(f'File not found: {fp}', file_path=str(fp))"));
        }
        if !self.SUPPORTED_FORMATS.contains(&fp.extension().unwrap_or_default().to_str().unwrap_or("").to_lowercase()) {
            return Err(anyhow::anyhow!("DocumentError(f'Unsupported format: {fp.extension().unwrap_or_default().to_str().unwrap_or("")}', file_path=str(fp), format=fp.extension().unwrap_or_default().to_str().unwrap_or(""))"));
        }
        if fp.stat().st_size > self.MAX_FILE_SIZE {
            return Err(anyhow::anyhow!("ValidationError(f'File too large ({fp.stat().st_size} bytes)', field='file_size')"));
        }
        // try:
        {
            let mut text = self.extract_text(fp.to_string());
            self.index_text(text, /* source= */ fp.name, /* metadata= */ HashMap::from([("original_path".to_string(), fp.to_string()), ("file_format".to_string(), fp.extension().unwrap_or_default().to_str().unwrap_or(""))]))
        }
        // except (ValidationError, DocumentError) as _e:
        // except Exception as exc:
    }
    /// Extract text from a file.
    pub fn extract_text(&mut self, file_path: String) -> Result<String> {
        // Extract text from a file.
        let mut fp = PathBuf::from(file_path);
        let mut suffix = fp.extension().unwrap_or_default().to_str().unwrap_or("").to_lowercase();
        // try:
        {
            if suffix == ".pdf".to_string() {
                self._extract_pdf(fp)
            }
            fp.read_to_string(), /* errors= */ "ignore".to_string())
        }
        // except Exception as exc:
    }
    /// List all indexed documents.
    pub fn list_indexed_documents(&mut self) -> Vec<HashMap<String, Box<dyn std::any::Any>>> {
        // List all indexed documents.
        let mut all_meta = self._load_all_metadata();
        all_meta.values().into_iter().collect::<Vec<_>>()
    }
    pub fn _generate_id() -> String {
        /* uuid */ "00000000-0000-0000-0000-000000000000".to_string().to_string()[..8]
    }
    /// Extract text from PDF using PyMuPDF (fitz) if available.
    pub fn _extract_pdf(path: PathBuf) -> Result<String> {
        // Extract text from PDF using PyMuPDF (fitz) if available.
        // try:
        {
            // TODO: import fitz
            let mut doc = fitz.open(path.to_string());
            let mut pages = doc.iter().map(|page| page.get_text()).collect::<Vec<_>>();
            doc.close();
            pages.join(&"\n\n".to_string())
        }
        // except ImportError as _e:
    }
    pub fn _save_metadata(&mut self, doc_id: String, meta: HashMap<String, Box<dyn std::any::Any>>) -> () {
        let mut all_meta = self._load_all_metadata();
        all_meta[doc_id] = meta;
        self.metadata_filestd::fs::write(&serde_json::to_string(&all_meta).unwrap(), /* encoding= */ "utf-8".to_string());
    }
    pub fn _load_all_metadata(&mut self) -> Result<HashMap<String, Box<dyn std::any::Any>>> {
        if self.metadata_file.exists() {
            // try:
            {
                serde_json::from_str(&self.metadata_file.read_to_string())).unwrap()
            }
            // except Exception as _e:
        }
        Ok(HashMap::new())
    }
}
